go/pkg/ddapi/translate.go

```diff
@@ -0,0 +1,208 @@
+      "points": [
+        [
+          1610032230,
+          0
+        ],
+        ...
+      ],
+      "tags": [
+        "version:7.24.1"
+      ],
+      "host": "x1carb6",
+      "type": "rate",
+      "interval": 10
+    },
+    {
+      "metric": "datadog.dogstatsd.client.metrics",
+      "points": [
+    ...
+    }
+  ]
+}
+
+Note that each object in the `series` array is a time series fragment for a
+specific metric. Each metric is defined by its name and other meta data (think:
+labels). The time series fragment is comprised of one or multiple data points /
+samples.
+*/
+func TranslateDDSeriesJSON(doc []byte) ([]*prompb.TimeSeries, error) {
+	// Attempt to deserialize entire JSON document, using the type definitions
+	// above including the custom deserialization function
+	// ddPoint.UnmarshalJSON().
+	var sfragments ddSeriesFragments
+	jerr := json.Unmarshal(doc, &sfragments)
+	if jerr != nil {
+		return nil, fmt.Errorf("invalid JSON doc: %v", jerr)
+	}
+
+	// TODO: Consider preallocating `promTimeSeriesFragments`
+	//nolint:prealloc
+	var promTimeSeriesFragments []*prompb.TimeSeries
+
+	for _, fragment := range sfragments.Fragments {
+		// Build up label set as a map to ensure uniqueness of keys.
+		labels := map[string]string{
+			// A time series fragment corresponds to a specific metric with a
+			// name. Store this metric name in a reserved Prometheus label.
+			"__name__": sanitizeMetricName(fragment.Name),
+			"instance": fragment.Host,
+			"job":      "ddagent",
+			// "host": "host",  // do not set host, that's `instance` in the
+			// Prom world.
+			"device":           fragment.Device,
+			"type":             fragment.Type,
+			"source_type_name": fragment.SourceTypeName,
+		}
+
+		// Translate dd agent tags into label k/v pairs. Upon unexpected tag
+		// structure, log a warning but otherwise proceed.
+		for _, tag := range fragment.Tags {
+			t := strings.SplitN(tag, ":", 2)
+
+			if len(t) != 2 {
+				log.Warnf("Invalid tag %s for metric: %s", tag, fragment.Name)
+				continue
+			}
+
+			// Prefix the tag name so that the source of this label is known
+			// (and can be queried for, with guarantees) and so that it can't
+			// override an "important" label, such as "instance".
+			tname := "ddtag_" + sanitizeLabelName(t[0])
+			tvalue := t[1]
+			labels[tname] = tvalue
+		}
+
+		// Create slice from `labels` map, with values being of type
+		// prompb.Label. For `prompb.TimeSeries` construction below. Skip
+		// prompb.Label construction for empty values (for example,
+		// `fragment.Device` may be empty).
+		promLabelset := make([]*prompb.Label, 0, len(labels))
+		for k, v := range labels {
+			if len(v) == 0 {
+				continue
+			}
+
+			l := prompb.Label{
+				Name:  k,
+				Value: v,
+			}
+			promLabelset = append(promLabelset, &l)
+		}
+
+		// Inspiration from
+		// https://github.com/open-telemetry/opentelemetry-go-contrib/blob/v0.15.0/exporters/metric/cortex/cortex.go#L385
+
+		var promSamples []prompb.Sample
```

Comment:
This should be preallocated.
```suggestion
		promSamples := make([]prompb.Sample, 0, len(fragment.Points))
```

Reply:
Thank you! Commit upcoming. Interestingly, the linter didn't complain about this one (but it did complain about the one below, that you also found).